eth_block.go

```diff
@@ -0,0 +1,363 @@
+var _ node.Node = (*EthBlock)(nil)
+
+/*
+  INPUT
+*/
+
+// FromBlockRLP takes an RLP message representing
+// an ethereum block header or body (header, ommers and txs)
+// to return it as a set of IPLD nodes for further processing.
+func FromBlockRLP(r io.Reader) (*EthBlock, []*EthTx, []*EthTxTrie, error) {
+	// We may want to use this stream several times
+	rawdata, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Let's try to decode the received element as a block body
+	var decodedBlock types.Block
+	err = rlp.Decode(bytes.NewBuffer(rawdata), &decodedBlock)
+	if err != nil {
+		if err.Error()[:41] != "rlp: expected input list for types.Header" {
+			return nil, nil, nil, err
+		}
+
+		// Maybe it is just a header... (body sans ommers and txs)
+		var decodedHeader types.Header
+		err := rlp.Decode(bytes.NewBuffer(rawdata), &decodedHeader)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		// It was a header
+		return &EthBlock{
+			Header:  &decodedHeader,
+			cid:     rawdataToCid(MEthBlock, rawdata),
+			rawdata: rawdata,
+		}, nil, nil, nil
+	}
+
+	// This is a block body (header + ommers + txs)
+	// We'll extract the header bits here
+	headerRawData := getRLP(decodedBlock.Header())
+	ethBlock := &EthBlock{
+		Header:  decodedBlock.Header(),
+		cid:     rawdataToCid(MEthBlock, headerRawData),
+		rawdata: headerRawData,
+	}
+
+	// Process the found eth-tx objects
+	ethTxNodes, ethTxTrieNodes := processTransactions(decodedBlock.Transactions(),
+		decodedBlock.Header().TxHash[:])
+
+	return ethBlock, ethTxNodes, ethTxTrieNodes, nil
+}
+
+// FromBlockJSON takes the output of an ethereum client JSON API
+// (i.e. parity or geth) and returns a set of IPLD nodes.
+func FromBlockJSON(r io.Reader) (*EthBlock, []*EthTx, []*EthTxTrie, error) {
+	var obj objJSONBlock
+	dec := json.NewDecoder(r)
+	err := dec.Decode(&obj)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	headerRawData := getRLP(obj.Result.Header)
+	ethBlock := &EthBlock{
+		Header:  &obj.Result.Header,
+		cid:     rawdataToCid(MEthBlock, headerRawData),
+		rawdata: headerRawData,
+	}
+
+	// Process the found eth-tx objects
+	ethTxNodes, ethTxTrieNodes := processTransactions(obj.Result.Transactions,
+		obj.Result.Header.TxHash[:])
+
+	return ethBlock, ethTxNodes, ethTxTrieNodes, nil
+}
+
+// processTransactions will take the found transactions in a parsed block body
+// to return IPLD node slices for eth-tx and eth-tx-trie
+func processTransactions(txs []*types.Transaction, expectedTxRoot []byte) ([]*EthTx, []*EthTxTrie) {
+	var ethTxNodes []*EthTx
+	transactionTrie := newTxTrie()
+
+	for idx, tx := range txs {
+		ethTx := NewTx(tx) // Will panic if it finds an error while parsing a tx
+		ethTxNodes = append(ethTxNodes, ethTx)
+		transactionTrie.addTx(idx, ethTx.RawData())
+	}
+
+	ethTxTrieNodes := transactionTrie.getNodes()
+	if !bytes.Equal(transactionTrie.rootHash(), expectedTxRoot) {
```

Comment:
Then shouldn't this not panic (just return an error)? That is, panics are usually reserved for programmer errors.